# 1. Introduction

In this project, I will pull some data from the Qunadl API. Qaundl is currently the most widely used aggregator of financial market data.

Qaundl has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, I will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. I'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

# 2. Sourcing and Loading

**2a. Import relevant libraries**

In [197]:
import requests
import collections
import json

**2b. Preliminary data analysis**

In [198]:
r = requests.get('https://www.quandl.com/api/v3/datasets/FSE/EON_X?start_date=2020-09-01&end_date=2020-09-01&api_key=').json()
r

{'dataset': {'id': 36592406,
  'dataset_code': 'EON_X',
  'database_code': 'FSE',
  'name': 'E.on Se (EON_X)',
  'description': 'Stock Prices for E.on Se (2020-08-18) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE000ENAG999',
  'refreshed_at': '2020-09-01T22:16:16.150Z',
  'newest_available_date': '2020-09-01',
  'oldest_available_date': '2003-01-20',
  'column_names': ['Date',
   'Open',
   'High',
   'Low',
   'Close',
   'Change',
   'Traded Volume',
   'Turnover',
   'Last Price of the Day',
   'Daily Traded Units',
   'Daily Turnover'],
  'frequency': 'daily',
  'type': 'Time Series',
  'premium': False,
  'limit': None,
  'transform': None,
  'column_index': None,
  'start_date': '2020-09-01',
  'end_date': '2020-09-01',
  'data': [['2020-09-01',
    9.95,
    10.045,
    9.89,
    9.89,
    None,
    5179826.0,
    51537357.58,
    None,
    None,
    None]],
  'collapse': None,
  'order': None,
  'database_id': 6129}}

In [199]:
for k in r.keys():
    print(k + ':', r[k])

dataset: {'id': 36592406, 'dataset_code': 'EON_X', 'database_code': 'FSE', 'name': 'E.on Se (EON_X)', 'description': 'Stock Prices for E.on Se (2020-08-18) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE000ENAG999', 'refreshed_at': '2020-09-01T22:16:16.150Z', 'newest_available_date': '2020-09-01', 'oldest_available_date': '2003-01-20', 'column_names': ['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover'], 'frequency': 'daily', 'type': 'Time Series', 'premium': False, 'limit': None, 'transform': None, 'column_index': None, 'start_date': '2020-09-01', 'end_date': '2020-09-01', 'data': [['2020-09-01', 9.95, 10.045, 9.89, 9.89, None, 5179826.0, 51537357.58, None, None, None]], 'collapse': None, 'order': None, 'database_id': 6129}


I will now conduct the following tasks on this dataset in order to answer some pertinent questions:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. What was the median trading volume during this year?

# 3. Business Insights

In [200]:
r2017 = requests.get('https://www.quandl.com/api/v3/datasets/FSE/EON_X?start_date=2017-01-01&end_date=2017-12-31&api_key=').json()

In [201]:
type(r2017)

dict

In [202]:
r2017

{'dataset': {'id': 36592406,
  'dataset_code': 'EON_X',
  'database_code': 'FSE',
  'name': 'E.on Se (EON_X)',
  'description': 'Stock Prices for E.on Se (2020-08-18) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE000ENAG999',
  'refreshed_at': '2020-09-01T22:16:16.150Z',
  'newest_available_date': '2020-09-01',
  'oldest_available_date': '2003-01-20',
  'column_names': ['Date',
   'Open',
   'High',
   'Low',
   'Close',
   'Change',
   'Traded Volume',
   'Turnover',
   'Last Price of the Day',
   'Daily Traded Units',
   'Daily Turnover'],
  'frequency': 'daily',
  'type': 'Time Series',
  'premium': False,
  'limit': None,
  'transform': None,
  'column_index': None,
  'start_date': '2017-01-01',
  'end_date': '2017-12-31',
  'data': [['2017-12-29',
    9.09,
    9.12,
    9.06,
    9.06,
    None,
    6460376.0,
    58678823.0,
    None,
    None,
    None],
   ['2017-12-28',
    9.15,
    9.16,
    9.06,
    9.08,
    None,
    5767048.0,
    52509

In [203]:
columns = r2017['dataset']['column_names']
columns

['Date',
 'Open',
 'High',
 'Low',
 'Close',
 'Change',
 'Traded Volume',
 'Turnover',
 'Last Price of the Day',
 'Daily Traded Units',
 'Daily Turnover']

In [204]:
columns.index('Open')

1

In [205]:
lowest_open = []
highest_open = []
for day in r2017['dataset']['data']:
    if lowest_open == [] and highest_open == []:
        lowest_open.append(day[1])
        highest_open.append(day[1])
    elif day[1] < lowest_open[0]:
        lowest_open[0] = day[1]
    elif day[1] > highest_open[0]:
        highest_open[0] = day[1]
print('Highest opening price in 2017: ' + str(highest_open[0]))
print('Lowest opening price in 2017: ' + str(lowest_open[0]))

Highest opening price in 2017: 10.75
Lowest opening price in 2017: 6.65


In [206]:
columns.index('High'), columns.index('Low')

(2, 3)

In [207]:
largest_change = []
for day in r2017['dataset']['data']:
    if largest_change == []:
        largest_change.append(day[2] - day[3])
    elif day[2] - day[3] > largest_change[0]:
        largest_change[0] = day[2] - day[3]
print('Largest change in any one day (based on High and Low price): ' + str(round(largest_change[0], 2)))

Largest change in any one day (based on High and Low price): 0.7


In [208]:
columns.index('Close')

4

In [209]:
keys = ['day1', 'day2']
closing = {key: None for key in keys}
largest_change_days = []
for day in r2017['dataset']['data']:
    closing['day1'] = closing['day2']
    closing['day2'] = day[4]
    if None in closing.values():
        largest_change_days.append(0)
    elif largest_change_days[0] < abs(closing['day2'] - closing['day1']):
        largest_change_days[0] = abs(closing['day2'] - closing['day1'])
print('Largest change between any two days (based on Closing price): ' + str(round(largest_change_days[0], 2)))

Largest change between any two days (based on Closing price): 0.48


In [210]:
columns.index('Traded Volume')

6

In [211]:
tv2017 = []
for day in r2017['dataset']['data']:
    tv2017.append(day[6])
tv2017_avg = sum(tv2017) / len(tv2017)
print('Average daily trading volume in 2017: ' + str('{:,}'.format(round(tv2017_avg, 2))))

Average daily trading volume in 2017: 11,708,661.14


In [212]:
def median(list):
    s = sorted(list)
    if len(s) % 2 == 0:
        median = (s[len(s)//2] + s[len(s)//2 + 1]) / 2
    else:
        median = s[len(s)//2 + 1]
    return '{:,}'.format(round(median, 2))

In [213]:
print('Median trading volume in 2017: ' + str(median(tv2017)))

Median trading volume in 2017: 10,561,328.5


# 4. Conclusion

Based on an analysis of Carl Zeiss Meditec's stock prices on the Frankfurt Stock Exchange, I was able to answer the following questions:

What was the largest change in any one day in 2017 (based on High and Low price)? **\$0.07**<br>
What was the largest change between any two days in 2017 (based on Closing Price)? **\$0.48**<br>
What was the average daily trading volume during this year? **\$11,708,661.14**<br>
What was the median trading volume during this year? **\$10,561,328.50**